### PR TITLE
Skip snapshot

### PR DESCRIPTION
**Pull request by AdalbertMemSQL**
Added ability to skip data snapshotting
Added ability to provide offsets
Refactored streaming to simplify it and to avoid deadlocks

### CODE REVIEW

File: src/main/java/com/singlestore/debezium/SingleStoreConnector.java [hunk 1]
```diff
@@ -115,4 +117,23 @@ public List<TableId> getMatchingCollections(Configuration config) {
...
+  @Override
+  public Config validate(Map<String, String> connectorConfigs) {
+    Config res = super.validate(connectorConfigs);
+
+    Configuration config = Configuration.from(connectorConfigs);
+    SingleStoreConnectorConfig connectorConfig = new SingleStoreConnectorConfig(config);
+    if (connectorConfig.getSnapshotMode() == SnapshotMode.SCHEMA_ONLY
+        && connectorConfig.offsets() == null) {
+      res.configValues().forEach(value -> {
+        if (value.name().equals(SingleStoreConnectorConfig.OFFSETS.name())) {
```
Comment (oyeliseiev-ua) [inline src/main/java/com/singlestore/debezium/SingleStoreConnector.java:130]: Would be great move this validation logic to SingleStoreConnectorConfig.OFFSETS.withValidation method.

File: src/main/java/com/singlestore/debezium/SingleStoreSnapshotChangeEventSource.java [hunk 2]
```diff
@@ -122,6 +122,12 @@ public SnapshotResult<SingleStoreOffsetContext> doExecute(ChangeEventSourceConte
       } else {
         LOGGER.info("Snapshot step 5 - Skipping snapshotting of data");
         releaseDataSnapshotLocks(ctx);
+        for (int i = 0; i < connectorConfig.offsets().size(); i++) {
+          String offset = connectorConfig.offsets().get(i);
+          if (!offset.equalsIgnoreCase("null")) {
+            ctx.offset.update(i, "", offset);
```
Comment (oyeliseiev-ua) [inline src/main/java/com/singlestore/debezium/SingleStoreSnapshotChangeEventSource.java:128]: How does it affect the streaming?

File: src/test/java/com/singlestore/debezium/StreamingIT.java [hunk 3]
```diff
@@ -627,4 +627,52 @@ public void testStaleOffsetInWhenNeededSnapshotMode() throws Exception {
...
+      start(SingleStoreConnector.class, config);
+      waitForStreamingToStart();
+      assertConnectorIsRunning();
+
+      records = consumeRecordsByTopic(1).allRecordsInOrder();
+      assertEquals(1, records.size());
+
+      record = records.get(0);
+      value = (Struct) record.value();
+      assertEquals(Integer.valueOf(2), value.getStruct("after").getInt32("a"));
```
Comment (oyeliseiev-ua) [inline src/test/java/com/singlestore/debezium/StreamingIT.java:673]: check also that it's streaming type record

Comment (oyeliseiev-ua) [reply src/test/java/com/singlestore/debezium/StreamingIT.java:673]: I would add also 1 more test for validator

**Review by oyeliseiev-ua — APPROVED** [review]

File: src/main/java/com/singlestore/debezium/SingleStoreConnector.java (hunk 1 above)
Comment (AdalbertMemSQL) [reply src/main/java/com/singlestore/debezium/SingleStoreConnector.java:130]: Nice idea. Will try to implement this.

File: src/main/java/com/singlestore/debezium/SingleStoreSnapshotChangeEventSource.java (hunk 2 above)
Comment (AdalbertMemSQL) [reply src/main/java/com/singlestore/debezium/SingleStoreSnapshotChangeEventSource.java:128]: Here we change the offset context which will be used when streaming is started.

File: src/test/java/com/singlestore/debezium/StreamingIT.java (hunk 3 above)
Comment (AdalbertMemSQL) [reply src/test/java/com/singlestore/debezium/StreamingIT.java:673]: Added

**Pull request merged**